db/version_set.cc

```diff
@@ -4226,15 +4228,33 @@ Status VersionSet::ProcessManifestWrites(
       // Each version in versions corresponds to a column family.
       // For each column family, update its log number indicating that logs
       // with number smaller than this should be ignored.
+      // TODO (yanqin): remove the nested loop if possible.
       for (const auto version : versions) {
         uint64_t max_log_number_in_batch = 0;
         uint32_t cf_id = version->cfd_->GetID();
+        std::string full_history_ts_low;
         for (const auto& e : batch_edits) {
-          if (e->has_log_number_ && e->column_family_ == cf_id) {
-            max_log_number_in_batch =
-                std::max(max_log_number_in_batch, e->log_number_);
+          if (e->column_family_ == cf_id) {
+            if (e->has_log_number_) {
+              max_log_number_in_batch =
+                  std::max(max_log_number_in_batch, e->log_number_);
+            }
+            if (e->HasFullHistoryTsLow()) {
+              assert(version->cfd_);
+              const Comparator* const ucmp = version->cfd_->user_comparator();
+              assert(ucmp);
+              if (full_history_ts_low.empty() ||
+                  ucmp->CompareTimestamp(full_history_ts_low,
+                                         e->GetFullHistoryTsLow()) <= 0) {
+                full_history_ts_low.assign(e->GetFullHistoryTsLow());
+              }
+            }
```

Comment:
I think you could just call `SetFullHistoryTsLow` directly on the `ColumnFamilyData` here instead of accumulating the timestamps in a local variable.